### ISSUE TITLE
Request.keep, drop for abstract interfaces

### ISSUE THREAD

**Issue by uNetworkingAB** [uNetworking/uWebSockets.js#1036]
There are ways to easily support more abstract interfaces such as the fetch one, without any significant perf. loss in fast path.

Request is stack allocated and is only valid for the "sync" callback. Adding a call, Request.keep() that would dynamically allocate it and make it "keepable" could be automatically done for callbacks that do not return a Response.

This means, in the fast path, there is no overhead, while in the slow async path, you can still easily "keep" the Request for as long as you need as a slow path.

The benefit here is obvious: wrappers that implement more abstract interfaces do not need to:

* Loop over all headers
* Create JavaScript string for all headers
* Hand over this object regardless if it will be kept or not

This is really bad for fast path performance

Internally, all that is neeed is adding 2 functions: getMemory, setMemory so that keep() calls malloc, getMemory, memcpy, setMemory on the Request and drop calls free

**Comment by uNetworkingAB** [comment]
Internally it should be

uWS::HttpRequest* = uWS::HttpRequest*->makeDynamicClone();
uWS::HttpRequest*->freeDynamicClone();

it just makes 1 block with the headers and all

**Comment by uasan** [comment]
Those. if we call `request.keep()`, we can access the request object, in the following microtasks?

**Comment by uNetworkingAB** [comment]
A wrapper is:

function myFetch(request) {
return new Response("hello");
}

app.get('\*', (res, req) => {
if myFetch(req) returned Response then
res.end(the response)
else if returned void or unresolved promise
req.keep(); // moves uWS.HttpRequest from global to local, swaps the underlying ptr to makeDynamicCopy(), on GC calls freeDynamicCopy()
endif
});

**Comment by uNetworkingAB** [comment]
Yes if you call keep() it would move from being a stack allocated thing to being dynamic and available until it gets GC:d. So it's a very simple function and wrappers would not need to copy out all headers one by done but can just keep the Request after calling keep() on it

**Comment by uNetworkingAB** [comment]
But it's slow as heck, so the key here is to only do it if the first invocation of the CB does not return a Response or calls end. So the whole idea is: only use it for slow path

**Comment by uasan** [comment]
I'm understood, thank you.

But the problem with all abstractions is that they usually do not know whether the request handler will be executed in the fast way or in the slow way, so the abstraction always chooses the slow way to be universal, so they will always call the keep method.

By what percentage does the response slow down if you always call the request method keep?

**Comment by uNetworkingAB** [comment]
> whether the request handler will be executed in the fast way or in the slow way

You don't need to know if it **will be**. You need to know if it *was*. Finding out if the request *was* ended is easy. That's the whole point of doing this *after* as a late stage slow path

**Comment by uNetworkingAB** [comment]
The library could even do this automatically, but it would add a cost to all async handlers. But those are kind of costly either way

**Comment by uNetworkingAB** [comment]
Making it automatic looks easiest. Then it would just work. And wrappers can remove all their header copying crap and always use the given Request object

**Comment by uasan** [comment]
Here's a more realistic example of abstraction:
```JS
async function myFetch(request) {
  return await sql;
}

app.get('*', async (response, request) => {
  try {
    const result = await myFetch(request);
    response.end(result);
  } catch (error) {
    response.end(error);
  }
});
```

> else if returned void or unresolved promise

There are misconceptions in this statement, the fact is that it is impossible to find out the state of a promise, whether it has resolved or not, such a promise interface, we can only wait for the result, so this is always the slow way and always calls the keep method

**Comment by uNetworkingAB** [comment]
There is no misconception. The library knows perfectly well whether the Response has ended or not. Like said, async/await functions will always take the slow path for a number of reasons (in Node.js, the microtask queue is not drained immediately after a call into JS which is terrible for perf. but I don't care it is a Node.js problem and doesn't affect non-async functions)

**Comment by uNetworkingAB** [comment]
Non-async functions like those in HelloWorld.js will end the request by calling Response.end() immediately and in this case there is no overhead since it never clones the request object

**Comment by uasan** [comment]
> The library knows perfectly well whether the Response has ended or not.

No, the point of abstractions is that the application code of the route handlers does not know anything about the implementation of the server, i.e. an asynchronous function that an abstraction calls never calls a response end, because this is the exclusive privilege of the abstraction, those. The response state after calling the handler will always be unfinished.

**Comment by uNetworkingAB** [comment]
"the library" = uWS. It knows whether you responded or not

**Comment by uasan** [comment]
Then the uWS should automatically call keep method for all asynchronous handlers, i.e. all asynchronous requests will be on the slow path, this will reduce the speed?

**Comment by porsager** [comment]
I think what you mean @uasan is that you'd like an option to "not" copy the request if you know you will not use anything from it in some of your async handlers?

**Comment by uNetworkingAB** [comment]
it could be negligible in comparison with async flow either way. needs benchmarking

**Comment by kartikk221** [comment]
This is a great proposal and would significantly help performance for wrappers / libraries like `hyper-express` where currently, the headers and other request properties are always consumed from `HttpRequest` and cached locally. With the ability to instruct `HttpRequest` to live past the first sync execution, we could convert the properties into getters which then lazy initialize the neccessary data and consume it as needed.

As for the best way of implementing it, automatic is ideal but having a method like `request.keep()` along with a `request.destroy()` would be fine as well since it is very easy to know If a request will be handled asynchronously.

**Comment by kartikk221** [comment]
> You don't need to know if it **will be**. You need to know if it _was_. Finding out if the request _was_ ended is easy. That's the whole point of doing this _after_ as a late stage slow path

I am asssuming to automate this behavior, you will be checking to see whether a response was sent after the route handler returns? I use a similar trick here to determine whether I need to cork before sending a response with a boolean flag:
https://github.com/kartikk221/hyper-express/blob/f712634ed5648b9c91d34beb0a845ab95ff3c6a8/src/components/Server.js#L526-L530

One more thing that I thought of is wouldn't someone calling `uWS.HttpResponse.onAborted(() => {})` be a strong indicator that they will handle the request in an asynchronous manner (slow path) hence keeping `uWS.HttpRequest` around in memory would be a safe bet?

**Comment by e3dio** [comment]
Probably the most difficult issue for devs using this lib is grabbing the request header items before doing any async. It's not hard but requires some planning. If the performance hit is small Auto-caching the request head for async would be a good default to improve DX. Then add an option like `App({ disableRequestCache })` that improves async perf with no heap allocation for people who know what they are doing. When `disableRequestCache` is enabled you could optionally call `request.keep()` to cache individual requests, but by default request caching is turned on so new devs don't even have to think about it. Needs benchmark to make sure performance is still good. Should also automatically call `onAborted(() => res.aborted = true)` for async responses, additional calls to `onAborted()` are optional. This cleans up the code and makes everything pretty easy

**Comment by e3dio** [comment]
> But it's slow as heck

Or maybe have the default be no request caching, add option `App({ enableRequestCache })` for all requests and `req.keep()` for individual

**Comment by Mupli** [comment]
Maybe we could have raw binary data only to improve at least caching mandatory things or some references/pointers etc.  

this.headerRawData = req.getHeaderRaw();

and when we really need them we could "unpack" data in the "lazy" manner on JS side.

**Comment by uNetworkingAB** [comment]
No, the interface of uWS is simple & efficient. If this feature is added it should be backwards compatible and seamless like discussed in the thread.

If there are "big" libraries that use uWS and would want this feature and collaborate in removing the copying and benchmarking it, I could consider it now.

**Comment by Mupli** [comment]
> If there are "big" libraries that use uWS and would want this feature and collaborate in removing the copying and benchmarking it, I could consider it now.

The big players doesn't care. The small guys with small VPS are the one that care and love the illogical performance improvements.

**Comment by mike-git374** [comment]
[hyper-express](https://github.com/kartikk221/hyper-express) from @kartikk221 was popular but hasn't had updates recently.

[ultimate-express](https://github.com/dimdenGD/ultimate-express) seems to be a new popular framework with recent updates and high TechEmpower [benchmarks](https://www.techempower.com/benchmarks/#section=data-r23&test=plaintext&l=zik0sf-pa7) from @dimdenGD and @nigrosimone maybe they would want to collaborate.

I'm using [node:sqlite](https://nodejs.org/docs/latest-v25.x/api/sqlite.html) so my stuff is synchronous but async could be helpful

**Comment by uNetworkingAB** [comment]
Great, that's a project that could give valuable input for this. If you support holding on to headers, you could potentially also support a more complex URL router that can support all kinds of things that these Express wrappers would want.

**Comment by uNetworkingAB** [comment]
You basically have 2 end user interfaces: Socket.IO and Express. For Socket.IO the "accelerator" is pub/sub support in uWS. For Express the "accelerator" is URL routing. Any such wrapper that makes use of those features would see big gains compared to having those in JS.

Since we already have pub/sub nailed down, making the URL router more like what these Express wrappers need _could_ be a side effect of this thread.

**Comment by dimdenGD** [comment]
(im ultimate-express creator) memory's hazy now, but i remember that getting headers made things slower. didnt have a way working around that, so this would be pretty nice. on unrelated note, for some reason reading ip (even raw one) is also a huge slowdown

**Comment by mustafa519** [comment]
I totally agree with @dimdenGD based on my experiences, getting the IP is also too expensive in uws somehow. I don't know why but I feel like there might be something wrong about it. Finally, someone mentioned about it.

**Comment by cesco69** [comment]
> Great, that's a project that could give valuable input for this. If you support holding on to headers, you could potentially also support a more complex URL router that can support all kinds of things that these Express wrappers would want.

In frameworks like ultimate-express, one of the most complex and performance-sensitive parts is the routing layer that tries to bridge uWebSockets.js with the Express-compatible public API.

Because uWS only exposes low-level routing, wrappers end up implementing fairly complex routers in JavaScript.

All of that introduces extra overhead and complexity that somewhat defeats the purpose of using uWS for maximum performance.

If uWS exposed a routing layer that was closer to what Express-style frameworks need (even if optional), wrappers like ultimate-express could offload a large part of their routing logic to native code instead of reimplementing it in JS.

**Comment by uNetworkingAB** [comment]
https://github.com/dimdenGD/ultimate-express looks to be the most practical and competently made Express support with uWS. And we can certainly take input from this project and consider changes but having Express-compatibility as goal/vision (in uWS itself) is not the right way to go.

I'm wondering - is it possible to get better results by following Fastify interfaces? Is Fastify designed after better algorithms?

**Comment by cesco69** [comment]
> I'm wondering - is it possible to get better results by following Fastify interfaces? Is Fastify designed after better algorithms?

That's a good question.

From my experience working on ultimate-express (I have made some PRs), the main bottleneck is not necessarily the Express interface itself, but the fact that the routing layer has to live partially in JavaScript.

For example, in ultimate-express we currently have three different routing tiers depending on what uWS can handle natively:

1. **Declarative responses**
These are the most efficient ones because the response is fully handled by uWS' declarative response API. However, they can only be used for static routes.

2. **Optimized routes**
These are routes where the path is compatible with the routing patterns supported by uWS, so we can register them directly in uWS and avoid additional routing logic.

3. **Non-optimized routes**
These are routes where the path pattern is not compatible with uWS routing, so they must be resolved by a JavaScript router before dispatching to the handler.

The third category is where most of the complexity and overhead comes from, because the router effectively runs outside of uWS.

For example the route response handler logic in ultimate-express ends up being quite involved because it needs to bridge these different cases:
https://github.com/dimdenGD/ultimate-express/blob/main/src/router.js#L466

Fastify-style APIs could help structurally (hooks instead of arbitrary middleware chains, flatter routing, etc.), but the main gains would probably come from allowing wrappers to delegate more of the routing work to uWS itself, especially pattern matching and parameter extraction.

That would reduce the need for JS routing layers and make frameworks on top of uWS much simpler and faster.

**Comment by cesco69** [comment]
Another area where wrappers end up adding a lot of complexity is streaming and backpressure handling.

When trying to provide a Node/Express-like response interface on top of uWS, frameworks need to bridge Node streams with the uWS backpressure model (tryEnd, onWritable, etc.). This often results in fairly complex logic just to correctly propagate backpressure and resume writes.

For example in ultimate-express the response streaming logic becomes quite involved because it needs to coordinate Node streams with the uWS response lifecycle:
https://github.com/dimdenGD/ultimate-express/blob/main/src/response.js#L133

If uWS exposed a higher-level public API for streaming responses (while still keeping the current low-level primitives), wrappers could implement Node-style streaming semantics with much less overhead and complexity.

This might also reduce the amount of glue code frameworks need to maintain around onWritable, buffering and resuming writes.
